### PR TITLE
Use stack optimizations.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/5914

Depends on https://github.com/ethereum/solidity/pull/5959

This still needs
 - [x] checks that it works with references to local valiables
 - [x] probably its own entry in the global optimizer settings
 - [x] changelog entry

### CODE REVIEW

File: docs/using-the-compiler.rst [hunk 1]
```diff
@@ -212,6 +212,9 @@ Input Description
             "cse": false,
             // Optimize representation of literal numbers and strings in code.
             "constantOptimizer": false,
+            // Improve allocation of stack slots for variables, can free up stack slots early.
+            // This can only be activated through the details here.
+            "stackAllocation": false,
```
Comment (axic) [inline docs/using-the-compiler.rst]: Isn't this a yul optimiser option?

Comment (chriseth) [reply docs/using-the-compiler.rst]: Yes it might be. It also influences inline assembly.

**Review by Marenz — DISMISSED** [review]

**Comment by chriseth** [comment]
@axic what about moving the setting to the details with a default of "true" if the yul optimizer is active? Should we display the details always as part of the metadata whenever the yul optimizer is activated?

**Comment by axic** [comment]
> what about moving the setting to the details with a default of "true" if the yul optimizer is active? 

If it cannot be selected separately I wouldn't include it as a setting. Just enable it if Yul is enabled?

> Should we display the details always as part of the metadata whenever the yul optimizer is activated?

I think we should only display options if they can be set on the input side.

**Comment by chriseth** [comment]
Hm, I think we have a misunderstanding there. I was proposing (input -> metadata output):

`details: { yul: true }` -> `details:{yul: true, yulDetails: {stackAllocation: true}}`

`details: { yul: true, yulDetails: {stackAllocation: false} }` -> `details:{yul: true, yulDetails: {stackAllocation: false}}`

`details: { yul: true, yulDetails: {stackAllocation: true} }` -> `details:{yul: true, yulDetails: {stackAllocation: true}}`


`details: { yul: false }` -> `details:{yul: false}`

While all other combinations are invalid for now.

**Comment by axic** [comment]
I think it is fine to put below `yulDetails` and turn it on by default.

File: libsolidity/interface/StandardCompiler.cpp [hunk 2]
```diff
@@ -409,12 +409,17 @@ boost::optional<Json::Value> StandardCompiler::parseOptimizerSettings(Json::Valu
...
+		if (settings.runYulOptimiser)
+			settings.optimizeStackAllocation = true;
 		if (details.isMember("yulDetails"))
 		{
-			if (!_jsonInput["yulDetails"].isObject())
-				return formatFatalError("JSONError", "The \"yulDetails\" optimizer setting has to be a JSON object.");
-			if (!_jsonInput["yulDetails"].getMemberNames().empty())
-				return formatFatalError("JSONError", "The \"yulDetails\" optimizer setting cannot have any settings yet.");
+			if (!details.isMember("yul"))
+				return formatFatalError("JSONError", "\"Providing yulDetails requires Yul optimizer to be enabled.");
```
Comment (ekpyron) [inline libsolidity/interface/StandardCompiler.cpp:407]: But the condition will also be true if the yul member is set to false, won't it? Shouldn't we either require the *value* to be in fact true or remove the check entirely?

**Review by ekpyron — COMMENTED** [review]
Looks fine apart from the one check.

File: libsolidity/interface/StandardCompiler.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/interface/StandardCompiler.cpp:407]: if yul is set to false, then settings.optimizeStackAllocation should be force to false, too...

Comment (ekpyron) [reply libsolidity/interface/StandardCompiler.cpp:407]: My point is that
```
  yulDetails: {}
```
will result in the error "Providing yulDetails requires Yul optimizer to be enabled."
Whereas:
```
  yul: false,
  yulDetails: {}
```
Will not error out, which is inconsistent.

Or am I missing something?

Comment (chriseth) [reply libsolidity/interface/StandardCompiler.cpp:407]: Ah, now I understand! So this should be changed to `if (!details.isMember("yul") || details["yul"].asBool() != true)`?

Comment (ekpyron) [reply libsolidity/interface/StandardCompiler.cpp:407]: Yep.

Comment (ekpyron) [reply libsolidity/interface/StandardCompiler.cpp:407]: Or maybe better and simpler: ``if (!settings.runYulOptimiser)``.

**Comment by chriseth** [comment]
Updated.

**Review by ekpyron — DISMISSED** [review]

File: libsolidity/interface/StandardCompiler.cpp [hunk 3]
```diff
@@ -399,12 +399,17 @@ boost::variant<OptimiserSettings, Json::Value> parseOptimizerSettings(Json::Valu
...
 		{
-			if (!_jsonInput["yulDetails"].isObject())
-				return formatFatalError("JSONError", "The \"yulDetails\" optimizer setting has to be a JSON object.");
-			if (!_jsonInput["yulDetails"].getMemberNames().empty())
-				return formatFatalError("JSONError", "The \"yulDetails\" optimizer setting cannot have any settings yet.");
+			if (!settings.runYulOptimiser)
+				return formatFatalError("JSONError", "\"Providing yulDetails requires Yul optimizer to be enabled.");
+
+			if (auto result = checkKeys(details["yulDetails"], {"stackAllocation"}, "settings.optimizer.details.yulDetails"))
+				return result;
```
Comment (ekpyron) [inline libsolidity/interface/StandardCompiler.cpp]: Weird - Did this compile before or somewhere as it is?
```suggestion
				return *result;
```

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: libsolidity/interface/StandardCompiler.cpp (hunk 3 above)
Comment (chriseth) [reply libsolidity/interface/StandardCompiler.cpp]: I think this was introduced with the rebase after refactoring.

**Comment by chriseth** [comment]
Fixed.

**Review by ekpyron — DISMISSED** [review]

**Comment by chriseth** [comment]
Actually I think this needs a test that the stack allocation setting is actually used. It is probably bad to do a bytecode comarison based one, but we could do a more clever check inside test/StandardCompiler.

We should also somehow test how the commandline interface passes the option.

And: we should check how this affects inline assembly.

**Comment by ekpyron** [comment]
One could probably construct an everywhere-immovable-expressions function that will throw the stack height error without the optimization and won't with the optimization and use that as a test.

**Comment by chriseth** [comment]
It turned out, the flag was actually not properly passed along from the commandline interface.

File: test/cmdlineTests.sh [hunk 4]
```diff
@@ -125,6 +125,8 @@ function test_solc_behaviour()
         sed -i -e '/^Warning: This is a pre-release compiler version, please do not use it in production./d' "$stderr_path"
         sed -i -e 's/ Consider adding "pragma .*$//' "$stderr_path"
     fi
+    # Remove path name to cpp file
```
Comment (erak) [inline test/cmdlineTests.sh]:
```suggestion
    # Remove path to cpp file
```

**Review by erak — DISMISSED** [review]

**Comment by chriseth** [comment]
Updated.

**Review by erak — APPROVED** [review]

**Pull request merged**